XLS-47d-PriceOracles/README.md

```diff
@@ -0,0 +1,468 @@
+              "AssetPrice": 74,
+              "Scale": 2,
+            }
+          },
+        ],
+        "LastUpdateTime": 743609414,
+        "PreviousTxnID": "C53ECF838647FA5A4C780377025FEC7999AB4182590510CA461444B207AB74A9",
+        "PreviousTxnLgrSeq": 56865244
+    }
+
+## Transactions
+
+This proposal introduces several new transactions to allow for the creation, update, and deletion of the `PriceOracle` object.
+
+### Transaction for creating or updating `PriceOracle` instance
+
+We define a new transaction **OracleSet** for creating or updating a `PriceOracle` instance.  Before the transaction can be submitted to create a new `PriceOracle` instance, the Oracle Provider has to do the following:
+
+- Create or own the `Account` on the XRPL with sufficient XRP balance to meet the XRP reserve and the transaction fee requirements.
+- The Oracle Provider has to publish the `Account` account public key so that it can be used for verification by dApp’s.
+- The Oracle Provider has to publish a registry of available Price Oracles with their unique Oracle Sequence. The hash of the `Account` and the `OracleSequence` uniquely identifies the Price Oracle on-ledger object.
+
+#### Transaction fields for **OracleSet** transaction
+
+|FieldName | Required? | JSON Type | Internal Type |
+|:---------|:-----------|:---------------|:------------|
+| `TransactionType` | :heavy_check_mark: | `string`| UINT16 |
+
+Indicates a new transaction type `OracleSet`. The integer value is TBD.
+
+|FieldName | Required? | JSON Type | Internal Type |
+|:---------|:-----------|:---------------|:------------|
+| `Account` | :heavy_check_mark: | `string` | `ACCOUNTID` |
+
+`Account` is the XRPL account that has update and delete privileges on the Oracle being set.
+
+|FieldName | Required? | JSON Type | Internal Type |
+|:---------|:-----------|:---------------|:------------|
+| `OracleSequence` | :heavy_check_mark: | `string` | `UINT32` |
+
+`OracleSequence` is a unique identifier of the Price Oracle for the given Account.
+
+|FieldName | Required? | JSON Type | Internal Type |
+|:---------|:-----------|:---------------|:------------|
+| `Provider` | | `string` | `BLOB` |
+
+`Provider` identifies an Oracle Provider. `Provider` must be included when creating a new instance of `PriceOracle`.
+
+|FieldName | Required? | JSON Type | Internal Type |
+|:---------|:-----------|:---------------|:------------|
+| `URI` | | `string` | `BLOB` |
+
+`URI` is an optional field to reference the price data off-chain.
+
+|FieldName | Required? | JSON Type | Internal Type |
+|:---------|:-----------|:---------------|:------------|
+| `AssetClass` | | `string` | `BLOB` |
+
+`AssetClass` describes the asset's type.
+
+|FieldName | Required? | JSON Type | Internal Type |
+|:---------|:-----------|:---------------|:-----------|
+| `LastUpdateTime` | :heavy_check_mark: | `number` | `UINT32` |
+
+`LastUpdateTime` is the specific point in time when the data was last updated.
```

Comment:
same comment as above - how about using ISO8601-compliant string ?

Reply:
Thanks for bringing this up. After some discussion decided to change to Unix Time.

Reply:
This is a comment from @thejohnfreeman : https://github.com/XRPLF/rippled/pull/4789#issuecomment-1819874166

I find the [description](https://github.com/XRPLF/XRPL-Standards/blob/e0a27d1e09234f422db83420ff7dcd62d3a54edc/XLS-47d-PriceOracles/README.md#example-of-priceoracle-json) of the `OracleID` confusing. The spec says this:

> We compute the PriceOracle object ID, a.k.a., OracleID, as the SHA-512Half of the following values, concatenated in order:
>  - The Oracle space key (0x52)
>  - The Owner Account ID
>  - The Oracle Sequence. This field must be passed to the transactions and it describes a unique Price Oracle sequence for the given account.

- `OracleID` is not mentioned in the [description](https://github.com/XRPLF/XRPL-Standards/blob/e0a27d1e09234f422db83420ff7dcd62d3a54edc/XLS-47d-PriceOracles/README.md#the-priceoracle-object) of the fields of `PriceOracle`, but it is [included](https://github.com/XRPLF/XRPL-Standards/blob/e0a27d1e09234f422db83420ff7dcd62d3a54edc/XLS-47d-PriceOracles/README.md#example-of-priceoracle-json) in the JSON of an example `PriceOracle`. Is it included in the ledger object or not?
- It is implied but not explicitly stated that the `OracleID` is used as the `SHAMap` key for a `PriceOracle` object. Is it? How are these objects supposed to be laid out in the ledger? The aggregation transaction suggests that it is possible to iterate `PriceOracle` objects for a given account. How?
- Oracle Sequence is not mentioned in the fields of `PriceOracle` or shown in the example. Is it possible to recover the `OracleID` for a `PriceOracle` using only the fields contained within a `PriceOracle`? If so, how?
- "Sequence" implies to me that it is a number that increments by one with each change, like a Ledger Sequence or an Account Sequence, but `OracleID` implies a constant identifier. Which is it? Are there any rules on the sequence number?

In the spec, the [list](https://github.com/XRPLF/XRPL-Standards/blob/e0a27d1e09234f422db83420ff7dcd62d3a54edc/XLS-47d-PriceOracles/README.md#user-content-transaction-fields-for-oracleset-transaction) following "The transaction fails if" includes:

> - The `URI` field length exceeds 64 bytes.
> - The `Provider` field length exceeds 64 bytes. 

But the [section](https://github.com/XRPLF/XRPL-Standards/blob/e0a27d1e09234f422db83420ff7dcd62d3a54edc/XLS-47d-PriceOracles/README.md#user-content-the-priceoracle-object) describing the `PriceOracle` on-ledger object fields says:

> - `Provider` identifies an Oracle Provider. It can be URI or any data, for instance chainlink. It is a string of up to 256 ASCII hex encoded characters (0x20-0x7E).
> - `URI` is an optional [URI](https://datatracker.ietf.org/doc/html/rfc3986) field to reference price data off-chain. It is limited to 256 bytes.

Which is correct? Can we fix it in the spec?

I find this explanation unclear:

> If an object with the `OracleID` Object ID already exists then the new token pairs are added to the Oracle instance `PriceDataSeries` array. Note that the order of the token pairs in the `PriceDataSeries` array is not important since the token pair uniquely identifies location in the `PriceDataSeries` array of the `PriceOracle` object. Also note that not every token pair price has to be updated. I.e., even though the `PriceOracle` may define ten token pairs, `OracleSet` transaction may contain only one token pair price update. In this case the missing token pair will not include `AssetPrice` and `Scale` fields. `PreviousTxnID` can be used to find the last updated Price Data for this token pair.

"If an object with the ID exists, then the new token pairs are added..." I'm assuming "new token pairs" are ones that do not appear in the `PriceOracle`. If so, then why no mention of "old token pairs"? They should be updating what appears in the `PriceOracle`, correct?

"In this case the missing token pair..." What missing token pair? The token pairs that are in the object, but not the transaction? If yes, then what is meant by "[they] will not include `AssetPrice` and `Scale` fields"? If they are missing, won't they "not include" _every_ field?

I feel like there is an implied behavior for this transaction that is not well described by the existing language. Here is how I would describe that behavior. Is it correct?

> An `OracleSet` transaction uniquely identifies a `PriceOracle` object with its `Account` and `OracleSequence` fields. If such an object does not yet exist in the ledger, it is created. Otherwise, the existing object is updated. The `Provider`, `URI`, and `AssetClass` fields are copied directly from the transaction, if present. `Provider` and `AssetClass` must be included in the transaction if the object is being created.
>
> The `PriceDataSeries` of the transaction is copied to a newly created `PriceOracle` object, or updates an existing object, like so:
> - `PriceData` objects for (`BaseAsset`, `QuoteAsset`) token pairs that appear in the transaction but not the object are copied to the object.
> - `PriceData` objects for token pairs that appear in both the transaction and the object are overwritten in the object.
> - `PriceData` objects for token pairs that appear only in the object are left unchanged.
>
> The order of token pairs in the transaction is not important because the token pair uniquely identifies the location of the `PriceData` object in the `PriceDataSeries` array of the `PriceOracle` object.
>
> `LastUpdateTime`, `PreviousTxnID`, and `PreviousTxnLgrSeq` are set in the same manner as for an `AccountSet` transaction.
>
> The owner reserve of the account is updated according to the difference in the size of the `PriceDataSeries` before and after the transaction is applied: 0 for missing, 1 for 1 - 5 objects, 2 for 6 - 10 objects.

`AssetClass` is a required field on a `PriceOracle` object and an optional field in an `OracleSet` transaction, just like `Provider`, but its field description does not say that it must be included in the transaction when creating a new instance of `PriceOracle`, unlike `Provider`. It must be included in the transaction, correct?

A side note on formatting. I think the presentation in the section on `PriceOracle` fields is best: _one_ table with all the fields, followed by a bulleted list of field descriptions. Not like the transaction sections with separate tables (repeating the same header row) for each field. I think the example should go directly after or before the table, not at the end of the section.